pkg/cache/application.go

```diff
@@ -628,6 +647,21 @@ func (app *Application) handleReleaseAppAllocationEvent(event *fsm.Event) {
 			}
 		}
 	}
+	app.postAppReleased()
+}
+
+func (app *Application) postAppReleased() {
+	if app.GetApplicationState() != events.States().Application.Resuming {
+		return
+	}
+	for _, task := range app.taskMap {
+		if task.placeholder && task.GetTaskState() != events.States().Task.Completed {
```

Comment:
This check is not working, even after correcting the missing parts in the `states.go` file. At this point the tasks are not in Completed state but in Scheduling or Bound. I spent some time to understand why the transition is not made, but I have no idea so far.
Even if the transition would work, the state transition is handled in an async way, so we cannot guarantee, that at this point the task will have the updated state. 

Checking the pos's status is also not a good idea, because since the pod is gone, thee task has an old reference.

Maybe we can check if the the task has the TIMEOUT set as terminationType, since that is always set by the core. @yangwwei @manirajv06  what do you think?

Reply:
Hmm, need to understand why the tasks are not in Completed state. Since the `deletePod` will trigger a `CompleteTask` event and that should move the task state to `Completed`. Can you pls print all the task states in this case?

Reply:
The transition is processed, but later, since it is an async message, so in this place we cannot rely on the task state. In my test scenarios I had a job with 4 placeholders, and the task state transition was processed after all the asks and allocations were released, so the app will stuck in Resuming state, because at this point all the tasks has the old state. 

What about checking here the terminationType?

Reply:
I've made the change as it seem to be the only option for our testing purpose. But still not convinced completely. How does this check guarantees that task has been completed (may be already completed or going to complete) from shim side?

Reply:
hi @kingamarton @manirajv06 IMO, we should leverage the state machine transitions to handle this. The problem here is where to put the check for transiting the app out of `Resuming`. I think we can do the following:

When the pod got actually deleted, it will call `context.NotifyTaskComplete()`:  https://github.com/apache/incubator-yunikorn-k8shim/blob/18b722990b2bee5cdfaedc441d4c76da604bbd40/pkg/cache/context.go#L423-L431, here currently we only send an event to the task. We can also send an event `AppTaskCompleted` to the app

```
appEv := NewSimpleApplicationEvent(appID, events.AppTaskCompleted) 
dispatcher.Dispatch(appEv) 
```

and we handle 

1. `Resuming` -> `Resuming` on `AppTaskCompleted`, if not all tasks are Completed
2. `Resuming` -> `Running` on `AppTaskCompleted`, if all tasks are in Completed state

I think this will work, could you please see if this makes sense?

Reply:
Overall approach makes sense to me from place (where event gets triggered) perspective, but I am not really sure the necessity of new app event and its transitions.

Based on the recent changes in this PR, app would be moved from "resuming" to "running" through run application event in newly introduced method postAppReleased(). Now, we can completely remove this flow and its corresponding state transitions (release* events etc). Instead, we can fire run application event only when all tasks are placeholder and completed in NotfiyTaskComplete() method. Above all, this check will happen only app state is resuming. Also, I don't see a need to keep the app state in resuming itself as it won't change by any other events at this point. BTW, NotifyTaskComplete() is being called from update and delete pod. But, I don't think it matters.

Thoughts?

Reply:
I think both approaches can work, but I think the cleaner solution is what @yangwwei mentioned, to handle the check in the application code, its state machine. 
My problem with adding it to `NotifyTaskComplete` we will change the whole method's purpose and scope.